File: shoring/flywheel/src/main/java/com/panopset/flywheel/CommandMatcher.java
package com.panopset.flywheel;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import com.panopset.compat.Logop;

/**
 * Flywheel command matcher.
 */
final class CommandMatcher {

  /**
   * Given a List of commands, tie the quit commands to the MatchableCommands.
   *
   * @param commands
   *          Commands to add structure to.
   * @return New List of commands.
   * @throws FlywheelException
   *           Thrown if there is an unmatched quit.
   */
  protected static List<Command> matchQuitCommands(final List<Command> commands)
      throws FlywheelException {
    FlywheelException fwe = null; 
    final List<MatchableCommand> matchedCommandsForDebugging = new ArrayList<>();
    final List<Command> rtn = new ArrayList<>();
    final Deque<MatchableCommand> stack = new ArrayDeque<>();
    for (Command command : commands) {
      updateStack(stack, rtn, command);
      if (command instanceof MatchableCommand) {
        matchedCommandsForDebugging.add((MatchableCommand) command);
        if (stack.isEmpty()) {
          rtn.add(command);
        }
        stack.push((MatchableCommand) command);
      } else if (command instanceof CommandQuit) {
        if (stack.isEmpty()) {
          for (MatchableCommand mc : matchedCommandsForDebugging) {
            Logop.info(String.format("Succussfully matched with quit, before un-matched quit found: %s", mc.toString()));
          }
          fwe = new FlywheelException("Un-matched quit found, stopping.");
          break;
        }
        stack.pop();
      }
    }
    if (fwe != null) {
      Logop.error("Failed to process the following template commands:");
      for (Command cmd : commands) {
        Logop.info(cmd.toString());
      }
      throw fwe;
    }
    return rtn;
  }
  
  private static void updateStack(Deque<MatchableCommand> stack, List<Command> rtn, Command command ) {
    if (stack.isEmpty()) {
      if (!(command instanceof MatchableCommand)) {
        rtn.add(command);
      }
    } else {
      stack.peek().getCommands().add(command);
    }
  }

  private CommandMatcher() {
  }
}

File: code/shoring/flywheel/src/main/kotlin/com/panopset/flywheel/CommandMatcher.kt
package com.panopset.flywheel

import com.panopset.compat.Logz
import java.util.*

/**
 * Flywheel command matcher.
 */
internal object CommandMatcher {
    /**
     * Given a List of commands, tie the quit commands to the MatchableCommands.
     *
     * @param commands
     * Commands to add structure to.
     * @return New List of commands.
     * @throws FlywheelException
     * Thrown if there is an unmatched quit.
     */
    fun matchQuitCommands(commands: List<Command>): List<Command> {
        var fwe: FlywheelException? = null
        val matchedCommandsForDebugging: MutableList<MatchableCommand> = ArrayList()
        val rtn: MutableList<Command> = ArrayList()
        val stack: Deque<MatchableCommand> = ArrayDeque()
        for (command in commands) {
            updateStack(stack, rtn, command)
            if (command is MatchableCommand) {
                matchedCommandsForDebugging.add(command)
                if (stack.isEmpty()) {
                    rtn.add(command)
                }
                stack.push(command)
            } else if (command is CommandQuit) {
                if (stack.isEmpty()) {
                    for (mc in matchedCommandsForDebugging) {
                        Logz.info(
                            String.format(
                                "Succussfully matched with quit, before un-matched quit found: %s",
                                mc.toString()
                            )
                        )
                    }
                    fwe = FlywheelException("Un-matched quit found, stopping.")
                    break
                }
                stack.pop()
            }
        }
        if (fwe != null) {
            Logz.errorMsg("Failed to process the following template commands:")
            for (cmd in commands) {
                Logz.info(cmd.toString())
            }
            throw fwe
        }
        return rtn
    }

    private fun updateStack(stack: Deque<MatchableCommand>, rtn: MutableList<Command>, command: Command) {
        if (stack.isEmpty()) {
            if (command !is MatchableCommand) {
                rtn.add(command)
            }
        } else {
            stack.peek().commands.add(command)
        }
    }
}
